File: python/1609. Even Odd Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isEvenOddTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        que = [root]
        f = 1 
        while que:
            sz = len(que)
            pre = float('-inf') if f else float('inf')
            while sz:
                sz -= 1
                root = que.pop(0)
                val = root.val
                if f:
                    if val&1 == 0:
                        return False
                    if val <= pre:
                        return False
                else:
                    if val&1 == 1:
                        return False
                    if val >= pre:
                        return False
                pre = val
                if root.left:
                    que.append(root.left)
                if root.right:
                    que.append(root.right)
            f ^= 1

        return True
        

File: go/src/leetcode/1609. Even Odd Tree/1609. Even Odd Tree.go
/**
@author ZhengHao Lou
Date    2021/12/25
*/
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}
/**
https://leetcode-cn.com/problems/even-odd-tree/
 */
func isEvenOddTree(root *TreeNode) bool {
	var lv int
	que := []*TreeNode{root}
	for len(que) != 0 {
		l := len(que)
		var last int
		if lv&1 == 1 {
			last = 1e6 + 10
		}
		for i := 0; i < l; i++ {
			node := que[0]
			que = que[1:]
			x := node.Val
			if lv&1 == 0 {		// even level
				if x&1 == 0 || x <= last {
					return false
				}
			} else {		// odd level
				if x&1 == 1 || x >= last {
					return false
				}
			}
			last = x

			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
		}
		lv++
	}

	return true
}
